packages/eslint-plugin/src/rules/strict-enums.ts

```diff
@@ -0,0 +1,804 @@
+
+/**
+ * See the comment for `EnumKind`.
+ *
+ * This rule can safely ignore other kinds of types (and leave the validation in
+ * question to the TypeScript compiler).
+ */
+const IMPOSSIBLE_ENUM_TYPES =
+  ts.TypeFlags.BooleanLike |
+  ts.TypeFlags.NonPrimitive |
+  ts.TypeFlags.ESSymbolLike;
+
+const ALLOWED_TYPES_FOR_ANY_ENUM_ARGUMENT =
+  ts.TypeFlags.Any |
+  ts.TypeFlags.Unknown |
+  ts.TypeFlags.Number |
+  ts.TypeFlags.String;
+
+export type Options = [];
+export type MessageIds =
+  | 'incorrectIncrement'
+  | 'mismatchedAssignment'
+  | 'mismatchedComparison'
+  | 'mismatchedFunctionArgument';
+
+export default util.createRule<Options, MessageIds>({
+  name: 'strict-enums',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description: 'Disallow the usage of unsafe enum code patterns',
+      recommended: false,
+      requiresTypeChecking: true,
+    },
+    messages: {
+      incorrectIncrement: 'You cannot increment or decrement an enum type.',
+      mismatchedAssignment:
+        'The type of the enum assignment does not match the declared enum type of the variable.',
+      mismatchedComparison:
+        'The two things in the comparison do not have a shared enum type.',
+      mismatchedFunctionArgument:
+        'The {{ ordinal }} argument in the function call does not match the declared enum type of the function signature.',
+    },
+    schema: [],
+  },
+  defaultOptions: [],
+  create(context) {
+    const parserServices = util.getParserServices(context);
+    const typeChecker = parserServices.program.getTypeChecker();
+
+    // ----------------
+    // Helper functions
+    // ----------------
+
+    /**
+     * If passed an enum member, returns the type of the parent. Otherwise,
+     * returns itself.
+     *
+     * For example:
+     * - `Fruit` --> `Fruit`
+     * - `Fruit.Apple` --> `Fruit`
+     */
+    function getBaseEnumType(type: ts.Type): ts.Type {
+      const symbol = type.getSymbol();
+      if (symbol === undefined) {
+        return type;
+      }
+
+      if (!util.isSymbolFlagSet(symbol, ts.SymbolFlags.EnumMember)) {
+        return type;
+      }
+
+      const { valueDeclaration } = symbol;
+      if (valueDeclaration === undefined) {
+        return type;
+      }
+
+      const parentType = getTypeFromTSNode(valueDeclaration.parent);
+      if (parentType === undefined) {
+        return type;
+      }
+
+      return parentType;
+    }
+
+    /**
+     * See the comment for the `EnumKind` enum.
+     */
+    function getEnumKind(type: ts.Type): EnumKind {
+      if (type.isUnion()) {
+        throw new Error(
+          'The "getEnumKind" function is not meant to be used on unions. Use the "getEnumKinds" function instead.',
+        );
+      }
+
+      if (type.isIntersection()) {
+        throw new Error(
+          'The "getEnumKind" function is not meant to be used on intersections.',
+        );
+      }
+
```

Comment:
There's no need for these developer-facing checks. If the function is used improperly, a unit test should break.

```suggestion
```